File: source_go/internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/ebook-renamer/go/internal/types"
	"github.com/rs/zerolog/log"
)

// Scanner handles file scanning operations
type Scanner struct {
	RootPath string
	MaxDepth int
}

// New creates a new Scanner instance
func New(path string, maxDepth int) (*Scanner, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, err
	}

	info, err := os.Stat(absPath)
	if err != nil {
		return nil, err
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("path is not a directory: %s", path)
	}

	return &Scanner{
		RootPath: absPath,
		MaxDepth: maxDepth,
	}, nil
}

// Scan walks the directory tree and returns a list of interesting files
func (s *Scanner) Scan() ([]*types.FileInfo, error) {
	var files []*types.FileInfo

	err := filepath.Walk(s.RootPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Warn().Err(err).Str("path", path).Msg("Error accessing path")
			return nil // Continue walking
		}

		// Calculate depth
		relPath, err := filepath.Rel(s.RootPath, path)
		if err != nil {
			return nil
		}
		depth := len(strings.Split(relPath, string(os.PathSeparator)))
		if relPath == "." {
			depth = 0
		}

		if depth > s.MaxDepth {
			if info.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		// Skip directories and hidden files/system dirs
		if info.IsDir() {
			if s.shouldSkip(path) {
				return filepath.SkipDir
			}
			return nil
		}

		if s.shouldSkip(path) {
			return nil
		}

		// Create FileInfo
		fileInfo, err := s.createFileInfo(path, info)
		if err != nil {
			log.Warn().Err(err).Str("path", path).Msg("Error creating file info")
			return nil
		}

		if fileInfo != nil {
			files = append(files, fileInfo)
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	log.Debug().Int("count", len(files)).Msg("Scanner found files")
	return files, nil
}

func (s *Scanner) createFileInfo(path string, info os.FileInfo) (*types.FileInfo, error) {
	originalName := info.Name()
	size := uint64(info.Size())
	modifiedTime := info.ModTime()

	// Detect extension
	extension := ""
	if strings.HasSuffix(originalName, ".tar.gz") {
		extension = ".tar.gz"
	} else if strings.HasSuffix(originalName, ".download") {
		extension = ".download"
	} else if strings.HasSuffix(originalName, ".crdownload") {
		extension = ".crdownload"
	} else {
		extension = filepath.Ext(originalName)
	}

	isFailedDownload := strings.HasSuffix(originalName, ".download") || strings.HasSuffix(originalName, ".crdownload")

	// Only check size for PDF and EPUB files
	isEbook := extension == ".pdf" || extension == ".epub"
	isTooSmall := !isFailedDownload && isEbook && size < 1024 // Less than 1KB

	return &types.FileInfo{
		OriginalPath:     path,
		OriginalName:     originalName,
		Extension:        extension,
		Size:             size,
		ModifiedTime:     modifiedTime,
		IsFailedDownload: isFailedDownload,
		IsTooSmall:       isTooSmall,
		NewName:          nil,
		NewPath:          "", // Will be set during normalization
	}, nil
}

func (s *Scanner) shouldSkip(path string) bool {
	filename := filepath.Base(path)

	// Skip hidden files/folders
	if strings.HasPrefix(filename, ".") {
		return true
	}

	info, err := os.Stat(path)
	if err == nil && info.IsDir() {
		// Skip download folders
		if strings.HasSuffix(filename, ".download") || strings.HasSuffix(filename, ".crdownload") {
			return true
		}
	}

	// Skip known system directories
	skipDirs := []string{"Xcode", "node_modules", ".git", "__pycache__"}
	for _, d := range skipDirs {
		if filename == d {
			return true
		}
	}

	return false
}

File: source_py/ebook_renamer/scanner.py
"""
File scanning functionality for the ebook renamer.
"""

import os
from pathlib import Path
from typing import List

from .types import FileInfo


class Scanner:
    """Handles file scanning operations."""
    
    def __init__(self, root_path: str, max_depth: int):
        self.root_path = root_path
        self.max_depth = max_depth
    
    def scan(self) -> List[FileInfo]:
        """Scan the directory for files matching criteria."""
        files = []
        
        for root, dirs, filenames in os.walk(self.root_path):
            # Calculate depth
            relative_path = os.path.relpath(root, self.root_path)
            if relative_path == ".":
                depth = 0
            else:
                depth = relative_path.count(os.sep)
            
            # Skip if depth exceeds max_depth
            if depth > self.max_depth:
                # Don't recurse deeper
                dirs[:] = []
                continue
            
            # Skip hidden directories and system directories
            dirs[:] = [d for d in dirs if not self._should_skip_directory(d)]
            
            for filename in filenames:
                if not self._should_skip_file(filename):
                    file_path = os.path.join(root, filename)
                    try:
                        file_info = self._create_file_info(file_path)
                        if file_info:
                            files.append(file_info)
                    except (OSError, IOError):
                        # Skip files that can't be accessed
                        continue
        
        return files
    
    def _should_skip_directory(self, dirname: str) -> bool:
        """Determine if a directory should be skipped."""
        # Skip hidden directories
        if dirname.startswith("."):
            return True
        
        # Skip known system directories
        skip_dirs = {"Xcode", "node_modules", ".git", "__pycache__"}
        if dirname in skip_dirs:
            return True
        
        return False
    
    def _should_skip_file(self, filename: str) -> bool:
        """Determine if a file should be skipped."""
        # Skip hidden files
        if filename.startswith("."):
            return True
        
        return False
    
    def _create_file_info(self, file_path: str) -> FileInfo:
        """Create a FileInfo struct for the given path."""
        stat = os.stat(file_path)
        original_name = os.path.basename(file_path)
        
        # Detect extension (including .tar.gz)
        extension = self._detect_extension(original_name)
        
        # Detect failed downloads
        is_failed_download = (original_name.endswith(".download") or 
                             original_name.endswith(".crdownload"))
        
        # Check if file is too small (only for PDF and EPUB files)
        is_ebook = extension in {".pdf", ".epub"}
        is_too_small = (not is_failed_download and 
                       is_ebook and 
                       stat.st_size < 1024)  # Less than 1KB
        
        return FileInfo(
            original_path=file_path,
            original_name=original_name,
            extension=extension,
            size=stat.st_size,
            modified_time=stat.st_mtime,
            is_failed_download=is_failed_download,
            is_too_small=is_too_small,
            new_path=file_path,
        )
    
    def _detect_extension(self, filename: str) -> str:
        """Detect file extension, including special cases like .tar.gz."""
        if filename.endswith(".tar.gz"):
            return ".tar.gz"
        elif filename.endswith(".download"):
            return ".download"
        elif filename.endswith(".crdownload"):
            return ".crdownload"
        else:
            return Path(filename).suffix
